### PR TITLE
Added slides on C++20 ranges

### LINKED ISSUE #49: C++ 20 : add slides about ranges

### PR DESCRIPTION
**Pull request by sponce**
All comments have been addressed and this is now ready to be merged

Closes #49

### CODE REVIEW

File: talk/morelanguage/ranges.tex [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    \item A range is a group of items that you can iterator over
+    \item It provides \mintinline{cpp}{begin} and \mintinline{cpp}{end} functions
+    \item Ranges concepts are located in \mintinline{cpp}{std::ranges} namespace
+    \item All STL containers are ranges
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{Variations on ranges}
+    There are actually different type of ranges
+    \begin{itemize}
+    \item \mintinline{cpp}{input_range}, \mintinline{cpp}{output_range}, \mintinline{cpp}{bidirectional_range},  \mintinline{cpp}{random_access_range},  \mintinline{cpp}{contiguous_range}
```
Comment (bernhardmgruber) [inline talk/morelanguage/ranges.tex:16]: if you want to be complete you can add `forward_range`

File: talk/morelanguage/ranges.tex [hunk 2]
```diff
@@ -0,0 +1,93 @@
+\subsection[range]{Ranges}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Ranges}
+  \begin{block}{The range concept}
+    \begin{itemize}
+    \item A range is a group of items that you can iterator over
+    \item It provides \mintinline{cpp}{begin} and \mintinline{cpp}{end} functions
```
Comment (bernhardmgruber) [inline talk/morelanguage/ranges.tex:8]: Hmm, it does not need to provide `begin` and `end` in the sense of member functions. A range `r` needs to have `std::ranges::begin(r)` "work" and vice-versa for `end`. I wonder whether we should rephrase the sentence to something like:

```suggestion
    \item It must provide iterators to its beginning and end, obtainable via \mintinline{cpp}{std::range::begin} and \mintinline{cpp}{std::range::end}
```

File: talk/morelanguage/ranges.tex [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    \item mapping the different types of iterators
+    \end{itemize}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Views}
+  \begin{block}{The view concept}
+    \begin{itemize}
+    \item A \mintinline{cpp}{view} is something that you apply on a range and performs some operation
```
Comment (bernhardmgruber) [inline talk/morelanguage/ranges.tex:26]: What you describe here is actually a range adaptor. But the definition of a view is actually nebulous. See: https://en.cppreference.com/w/cpp/ranges/view
If I come across a better definition I will let you know! :)

On the other hand, range adapters create views.

File: talk/morelanguage/ranges.tex [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    \item A \mintinline{cpp}{view} does not own data
+    \item It's time to copy, move, assign is constant
+    \item \mintinline{cpp}{view}s can be applied with \mintinline{cpp}{|} and are lazy
+    \item They live in the \mintinline{cpp}{std::views} namespace
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{Example}
+    { \scriptsize
+      \begin{cppcode*}{gobble=4}
+        auto const ints = {0,1,2,3,4,5};
```
Comment (bernhardmgruber) [inline talk/morelanguage/ranges.tex:36]:
```suggestion
        auto const ints = std::views::iota(1, 6);
```

File: talk/morelanguage/ranges.tex [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                               | std::views::transform(square);
+        for (auto a : results) { ... } // 0, 4, 16
+      \end{cppcode*}
+    }
+  \end{exampleblock}
+\end{frame}
+
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Many \texttt{view}s are provided by the STL}
```
Comment (bernhardmgruber) [inline talk/morelanguage/ranges.tex:49]: Again, these are range adapters resulting in the corresponding views. E.g. `std::views::drop` is a range adapter, creating the `std::rangers::drop_view` view.

File: talk/morelanguage/ranges.tex [hunk 6]
```diff
@@ -0,0 +1,91 @@
+\subsection[range]{Ranges}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Ranges}
+  \begin{block}{The range concept}
+    \begin{itemize}
+    \item A range is a group of items that you can iterator over
+    \item It must provide iterators to its beginning and end, obtainable via \mintinline{cpp}{std::range::begin} and \mintinline{cpp}{std::range::end}
+    \item Ranges concepts are located in \mintinline{cpp}{std::ranges} namespace
```
Comment (bernhardmgruber) [inline talk/morelanguage/ranges.tex:9]:
```suggestion
    \item Range concepts are located in \mintinline{cpp}{std::ranges} namespace
```

File: talk/morelanguage/ranges.tex [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    \item All STL containers are ranges
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{Variations on ranges}
+    There are actually different type of ranges
+    \begin{itemize}
+    \item \mintinline{cpp}{input_range}, \mintinline{cpp}{output_range}, \mintinline{cpp}{bidirectional_range}, \mintinline{cpp}{random_access_range}, \mintinline{cpp}{contiguous_range}, \mintinline{cpp}{forward_range}
```
Comment (bernhardmgruber) [inline talk/morelanguage/ranges.tex:16]:
```suggestion
    \item \mintinline{cpp}{input_range}, \mintinline{cpp}{output_range}, \mintinline{cpp}{forward_range}, \mintinline{cpp}{bidirectional_range}, \mintinline{cpp}{random_access_range}, \mintinline{cpp}{contiguous_range}
```
Let's order them like the iterators.

File: talk/morelanguage/ranges.tex [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    \item mapping the different types of iterators
+    \end{itemize}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Views}
+  \begin{block}{The view concept}
+    \begin{itemize}
+    \item A \mintinline{cpp}{view} is a non owning object build from a range or another view after applying some range adaptor(s)
```
Comment (bernhardmgruber) [inline talk/morelanguage/ranges.tex:26]:
```suggestion
    \item A \mintinline{cpp}{view} is a non-owning object built from a range or another view after applying some range adaptor(s)
```

File: talk/morelanguage/ranges.tex [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    \item Its time to copy, move, assign is constant
+    \item Range adaptors can be applied with \mintinline{cpp}{|} and are lazy
```
Comment (bernhardmgruber) [inline talk/morelanguage/ranges.tex:28]:
```suggestion
    \item Range adaptors can be applied/changed with \mintinline{cpp}{|} and are lazy
```

**Review by bernhardmgruber — APPROVED** [review]
A few remaining suggestions, the rest LGTM!

File: talk/morelanguage/ranges.tex [hunk 10]
```diff
@@ -0,0 +1,91 @@
+\subsection[range]{Ranges}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Ranges}
+  \begin{block}{The range concept}
+    \begin{itemize}
+    \item A range is a group of items that you can iterator over
```
Comment (hageboeck) [inline talk/morelanguage/ranges.tex:7]:
```suggestion
    \item A range is a group of items that you can iterate over
```

File: talk/morelanguage/ranges.tex [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    \item It must provide iterators to its beginning and end, obtainable via \mintinline{cpp}{std::range::begin} and \mintinline{cpp}{std::range::end}
+    \item Ranges concepts are located in \mintinline{cpp}{std::ranges} namespace
+    \item All STL containers are ranges
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{Variations on ranges}
+    There are actually different type of ranges
+    \begin{itemize}
+    \item \mintinline{cpp}{input_range}, \mintinline{cpp}{output_range}, \mintinline{cpp}{bidirectional_range}, \mintinline{cpp}{random_access_range}, \mintinline{cpp}{contiguous_range}, \mintinline{cpp}{forward_range}
+    \item mapping the different types of iterators
```
Comment (hageboeck) [inline talk/morelanguage/ranges.tex:17]: I don't know if this is better:
```suggestion
    \item corresponding to the different iterator categories
```

File: talk/morelanguage/ranges.tex [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    \end{itemize}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Views}
+  \begin{block}{The view concept}
+    \begin{itemize}
+    \item A \mintinline{cpp}{view} is a non owning object build from a range or another view after applying some range adaptor(s)
+    \item Its time to copy, move, assign is constant
```
Comment (hageboeck) [inline talk/morelanguage/ranges.tex:27]:
```suggestion
    \item The time complexity to copy, move, assign is constant
```

File: talk/morelanguage/ranges.tex [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+      \begin{cppcode*}{gobble=4}
+        auto const ints = std::views::iota(0, 6);
+        auto even = [](int i) { return 0 == i % 2; };
+        auto square = [](int i) { return i * i; };
+        auto results = numbers | std::views::filter(even)
+                               | std::views::transform(square);
+        for (auto a : results) { ... } // 0, 4, 16
+      \end{cppcode*}
+     }
+  \end{block}
```
Comment (hageboeck) [inline talk/morelanguage/ranges.tex:90]: Why is this box identical to the example on the previous slide? Did you have in mind to mention something specific about this code?

**Review by hageboeck — COMMENTED** [review]
Looks quite good to me. I added a few small comments.

**Pull request merged**